### ISSUE TITLE
Virtual keyboard keymap source can be changed

### ISSUE THREAD

**Issue by dcz-purism** [swaywm/wlroots#2420]
This question comes from the issue found in phoc: https://source.puri.sm/Librem5/phoc/-/merge_requests/191

The problem is about who has control over the key map which is assigned to the virtual-keyboard device.

Quoting the issue:

> The authoritative source of keymap on virtual keyboards is the virtual-keyboard client. Changing its keymap causes the client to start sending invalid keycodes.

Because it's the virtual-keyboard client that supplies all the key codes, it should always know the current key map. The protocol doesn't provide any way to update that knowledge, it assumes that only the client ever updates the key map.

So when `wlr_keyboard_set_keymap` is called on the virtual keyboard (outside of wlroots itself), this breaks virtual-keyboard.

My question is: how do we go about preventing that?

Couple of ideas:

1. Leave it to the wlroots API consumer (compositor) to check that `wlr_input_device::name != "virtual keyboard"`
2. Same, but invent a well known `wlr_input_device::vendor, product`
3. Same, but create a `wlr_input_device::can_change_keymap` flag
4. Add a new function `wlr_keyboard_impl::set_keymap` and call it from `wlr_keyboard_set_keymap`

Notes: `wlr_keyboard_set_keymap` doesn't have access to `wlr_input_device`, so 1, 2, 3 cannot be done there. I also don't like how 1, 2, 3 create a contract that is not enforced by code, but relies on the API user being careful. 4. is the least bad, but if other protocols have other semantics in the future, `wlr_keyboard_impl` could become messy.

Would a patch based on 4. be accepted, or does a different solution sound better?

**Comment by emersion** [comment]
I agree compositors should somehow avoid switching the layout of a virtual keyboard.

> Leave it to the wlroots API consumer (compositor) to check that wlr_input_device::name != "virtual keyboard"

A more bulletproof way to do this would be to check `wlr_input_device_get_virtual_keyboard() != NULL`.

>Same, but create a wlr_input_device::can_change_keymap flag

We could check that flag from `wlr_keyboard_set_keymap` maybe?

>Add a new function wlr_keyboard_impl::set_keymap and call it from wlr_keyboard_set_keymap

This sounds like a large change for a little feature. I'm unclear if it's worth it.

>other protocols have other semantics in the future

We should maybe ask @whot how [libei](https://gitlab.freedesktop.org/libinput/libei) will handle this.

**Comment by dcz-purism** [comment]
> We could check that flag from wlr_keyboard_set_keymap maybe?

That sounds optimal, except `wlr_keyboard` doesn't carry enough data to check this: IDs are on `wlr_input_device`, which references `wlr_keyboard`, but not the other way around. Or do you propose extending `wlr_keyboard` with this information?

Side note: other virtual keyboard implementations may be more forgiving about keymap changes.

**Comment by whot** [comment]
> We should maybe ask @whot how libei will handle this.

The current libei API allows the client to set a keymap and the server implementation (i.e. the compositor) has to either accept, unset [1], or change that keymap to a different one. 

In the "accept" case it's up to the server to handle the rest, in the "unset" or "changed" case it's up to the client what to do next. A virtual keyboard would likely error out with "sorry, server does not accept our keymap so we cannot work correctly" or something.

[1] "unset" is equivalent to "you don't get a special keymap, you have to deal with whatever is applied by default to this device". Or, in less confusing terms: how any physical keyboard works right now, where the hw doesn't have any idea what a specific key code produces.

**Comment by dcz-purism** [comment]
I could implement the same semantic, but I would need to make a breaking protocol change: introduce a new "no, this is your key map instead" message. That would mean another unavoidable round trip when switching key maps (with Squeekboard this can happen now while typing Japanese in between characters).

I would prefer to embed the current semantics into how wlroots handles keyboards. On the other hand, virtual-keyboard is very unstable, so maybe we can get away with a protocol change and eat the round trip.
